### PR TITLE
Add MariaDB Scaling

### PR DESCRIPTION
**Pull request by Kidswiss**
## Summary

With this MariaDB can scale to a 3 node Galera cluster.
    
* Added ProxySQl statefulset
* Updated default version of MariaDB
* Updated Galera helm chart
* Added sensible PDBs to ProxySQL
## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.
- [ ] Link this PR to related issues.

### CODE REVIEW

File: apis/vshn/v1/dbaas_vshn_mariadb.go [hunk 1]
```diff
@@ -78,12 +83,22 @@ type VSHNMariaDBParameters struct {
...
 
 // VSHNMariaDBServiceSpec contains MariaDB DBaaS specific properties
 type VSHNMariaDBServiceSpec struct {
-	// +kubebuilder:validation:Enum="10.4";"10.5";"10.6";"10.9";"10.10";"10.11";"11.0";"11.1";"11.2";
-	// +kubebuilder:default="11.2"
+	// +kubebuilder:validation:Enum="10.4";"10.5";"10.6";"10.9";"10.10";"10.11";"11.0";"11.1";"11.2";"11.3";"11.4";"11.5";
+	// +kubebuilder:default="11.5"
 
 	// Version contains supported version of MariaDB.
 	// Multiple versions are supported. The latest version "11.2" is the default version.
```
Comment (zugao) [inline apis/vshn/v1/dbaas_vshn_mariadb.go:104]:
```suggestion
	// Multiple versions are supported. The latest version "11.5" is the default version.
```

File: pkg/comp-functions/functions/vshnmariadb/mariadb_deploy.go [hunk 2]
```diff
@@ -396,3 +411,64 @@ func getReleaseSecretValue(svc *runtime.ServiceRuntime, comp *vshnv1.VSHNMariaDB
 	}
 	return m["release"]
 }
+
+// To make scaling as seamless as possible we create an additional service.
+// This service will either point to a single db instance, or to the proxysql cluster.
+func createMainService(comp *vshnv1.VSHNMariaDB, svc *runtime.ServiceRuntime, secretName string) error {
```
Comment (zugao) [inline pkg/comp-functions/functions/vshnmariadb/mariadb_deploy.go:417]: Technically speaking the customer can also use the svc directly from galera cluster bypassing proxysql, is that right?

**Review by zugao — APPROVED** [review]
LGTM

The ProxySQL settings will be managed in a different PR right?

**Comment by Kidswiss** [comment]
> LGTM
> 
> The ProxySQL settings will be managed in a different PR right?

Do you mean the POC with the mysql client? Because right now all the settings are managed, just via the config file.

**Comment by zugao** [comment]
> Do you mean the POC with the mysql client? Because right now all the settings are managed, just via the config file.

Ah yes, we agreed to first start with config file, all good.

File: pkg/comp-functions/runtime/function_mgr.go [hunk 3]
```diff
@@ -482,6 +482,14 @@ func KubeOptionAddRefs(refs ...xkube.Reference) KubeObjectOption {
...
 }
 
+// KubeOptionAllowDeletion adds the allow deletion label to the object
+func KubeOptionAllowDeletion(obj *xkube.Object) {
+	if obj.ObjectMeta.Labels == nil {
+		obj.ObjectMeta.Labels = map[string]string{}
+	}
+	obj.ObjectMeta.Labels[WebhookAllowDeletionLabel] = "true"
+}
+
```
Comment (wejdross) [inline pkg/comp-functions/runtime/function_mgr.go:492]: why do we have this here? I can't find references in code

**Review by wejdross — APPROVED** [review]

File: pkg/comp-functions/runtime/function_mgr.go (hunk 3 above)
Comment (Kidswiss) [reply pkg/comp-functions/runtime/function_mgr.go:492]: It's used multiple times in `proxysql.go`:
https://github.com/vshn/appcat/pull/249/files#diff-6c00f19154044c901c94120e2e51fa40f7ccb6e422335f9172257d054447d96dR198
https://github.com/vshn/appcat/pull/249/files#diff-6c00f19154044c901c94120e2e51fa40f7ccb6e422335f9172257d054447d96dR258
https://github.com/vshn/appcat/pull/249/files#diff-6c00f19154044c901c94120e2e51fa40f7ccb6e422335f9172257d054447d96dR414
https://github.com/vshn/appcat/pull/249/files#diff-6c00f19154044c901c94120e2e51fa40f7ccb6e422335f9172257d054447d96dR476
https://github.com/vshn/appcat/pull/249/files#diff-6c00f19154044c901c94120e2e51fa40f7ccb6e422335f9172257d054447d96dR501

Comment (wejdross) [reply pkg/comp-functions/runtime/function_mgr.go:492]: file was hidden :see_no_evil: now it makes more sense :D

File: pkg/comp-functions/functions/vshnmariadb/mariadb_deploy.go (hunk 2 above)
Comment (Kidswiss) [reply pkg/comp-functions/functions/vshnmariadb/mariadb_deploy.go:418]: Yes, netpols only apply to pods and not the svcs... so there's no easy way to block this.
https://kubernetes.io/docs/concepts/services-networking/network-policies/#what-you-can-t-do-with-network-policies-at-least-not-yet

But on the other hand, it's probably only a question of time until we find the first app that doesn't like proxysql, just like with pgbouncer, so having a direct option for that might be a good thing.

File: pkg/comp-functions/functions/vshnmariadb/proxysql.go [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+        {writer_hostgroup=2,backup_writer_hostgroup=4,reader_hostgroup=3,offline_hostgroup=1,active=1,max_writers=1,writer_is_also_reader=1,max_transactions_behind=100}
+    )
+
+    proxysql_servers =
+    (
+        { hostname = "proxysql-0.proxysqlcluster.{{ .Namespace }}.svc", port = 6032, weight = 1 },
+        { hostname = "proxysql-1.proxysqlcluster.{{ .Namespace }}.svc", port = 6032, weight = 1 }
+    )`
+)
+
```
Comment (wejdross) [inline pkg/comp-functions/functions/vshnmariadb/proxysql.go:69]: can't we just move it to separate file and then template it with text/template? 
#readibility

**Review by wejdross — APPROVED** [review]

File: pkg/comp-functions/functions/vshnmariadb/proxysql.go (hunk 4 above)
Comment (Kidswiss) [reply pkg/comp-functions/functions/vshnmariadb/proxysql.go:69]: Good idea!

**Review by zugao — APPROVED** [review]
LGTM

**Pull request merged**